man/buffers.3
.TH BUFFERS 3 "12 September 1995"
.FB
.SH NAME
w_init, w_write, w_writen, nextwrite, w_flush \- general file buffering
interface
.SH SYNOPSIS
.nf
.nj
fb_w_init(n, fd0, fd1)
int n, fd0, fd1;

fb_w_write(ch, buf)
int ch;
char *buf;

fb_w_writen(ch, buf, n)
int ch, n;
char *buf;

fb_nextwrite(ch, buf)
int ch;
char *buf;

fb_w_flush(n)
int n;
.fi
.ju
.PP
.SH DESCRIPTION
The buffers package is used to write large amounts of text to one or two files
in a buffered, disk efficient manner. Without changing the logic of
normal algorithms that often write to files in pieces,
the buffers package enforces the blocking of bytes into single 1024-byte
disk accesses.
.PP
The files being written to need to be opened before using the buffers package.
Also, file descriptors are used, not file streams. Again, integer file
descriptors, like those returned by \fIopen\fP(2) are used inside the buffers
package.
.PP
The buffers package can handle two different output files. Each is assigned
a channel within the buffers package. This channel number is then used
by the outside layers when writing to a file descriptor.
.PP
To initialize the buffers package, \fIfb_w_init\fP is used.
The argument \fIn\fP,
set to either one or two, determines how may channels will be used; \fIfd0\fP
is for channel 0 (zero), \fIfd1\fP is for channel 1 (one).
When all completed with
the buffers, use of \fIfb_w_flush\fP will flush the "written" text, forcing any
remnants to the disk. The argument to \fIfb_w_flush,\fP
\fIn\fP, is the number of
channels.
.PP
There are three routines for depositing bytes to be written by the buffers
package: \fIfb_w_write\fP, \fIfb_w_writen\fP, and \fIfb_nextwrite\fP.
The first of these, \fIfb_w_write\fP, writes a single byte to channel \fIch\fP.
The byte written is pointed to by \fIbuf\fP (yes, a character pointer).
The next routine, \fIfb_w_writen\fP is almost the same as \fIfb_w_write\fP, but
it writes \fIn\fP bytes from \fIbuf\fP.
.PP
The last routine, \fIfb_nextwrite\fP, is used to write strings to a buffers
channel. This routine will write each byte up to but not including the
NULL terminator at the end of \fIbuf\fP.
.PP
These routines together are used at times when a data file can be
written in large (1024 byte) chunks. Use of this buffers package can vastly
increase the speed of some processes.
.PP
NOTE: the buffers package is usually not needed unless you
are generating FirstBase databases using buffered writes (see
$FIRSTBASEHOME/applications/sample_src/bload.c) or something similar. 
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/libcc/buffers.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: buffers.c,v 9.0 2001/01/09 02:56:19 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Buffers_sid[] = "@(#) $Id: buffers.c,v 9.0 2001/01/09 02:56:19 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

#define BLKSIZE	1024
#define NWRITES	3
#define MAXLINE 250

static char r_buffer[BLKSIZE + 1];	/* for buffered read */
static int r_curp = -1;			/* current pointer into buffer */
static int r_maxp = 0;			/* max pointer */
static int r_fd;			/* read file descriptor */

static char *w_buffer[NWRITES];		/* for buffered write buffers */
static int w_curp[NWRITES];		/* for buffer current pointers */
static int w_fd[NWRITES];		/* for file descriptors */

/*
 * r_init - initialize buffered read system
 */
   fb_r_init(fd)
      int fd;

      {
         r_curp = -1;
	 r_fd = fd;
      }

/*
 * r_end - end buffered read system
 */
   fb_r_end()

      {
	 close(r_fd);
      }

/*
 * r_rewind - re_initialize buffered read system
 */
   fb_r_rewind()

      {
	 if (lseek(r_fd, 0L, 0) != 0)
            fb_xerror(FB_SEEK_ERROR, SYSMSG[S_BAD_DATA], NIL);
         r_curp = -1;
      }

/*
 * nextread - semi buffered read - read a block if needed.
 */
   fb_nextread(buf)
      char *buf;
      
      {
         if (r_curp < 0 || r_curp >= r_maxp){
	    r_maxp = read(r_fd, r_buffer, BLKSIZE);
	    if (r_maxp <= 0)
	       return(0);
	    r_curp = 0;
	    }
	 *buf = r_buffer[r_curp++];
         return(1);
      }
 
/*
 * nextline - get a line from input, strip NEWLINE, and store NULL
 */

   fb_nextline(buf, max)
      char *buf;
      int max;

      {
         char c;
	 int i;

         if (max <= 0)
            max = MAXLINE;
	 for (i = 1; i < max; i++){		/* 1..MAXLINE inclusive */
	    if (fb_nextread(&c) == 0)
	       return(0);
	    if (c == '\n')
	       break;
	    *buf++ = c;
	    }
	 *buf = 0;
	 return(i);
      }

/*
 * w_init - initialize N buffered write systems, fds of a ... z
 */
   fb_w_init(n, a, b)
      int n, a, b;

      {
         int i;
	 extern char *fb_malloc(unsigned s);

	 w_fd[0] = a; w_fd[1] = b;
         for (i = 0; i < n; i++){
	    w_curp[i] = 0;
	    w_buffer[i] = (char *) fb_malloc(BLKSIZE + 1);
	    }
      }

/*
 * w_end - end N buffered write systems
 */
   fb_w_end(n)
      int n;

      {
         int i;

         for (i = 0; i < n; i++)
	    close(w_fd[i]);
      }

/*
 * nextwrite - buffered write - write string. check for a needed flush.
 */
   fb_nextwrite(ch, buf)
      int ch;
      char *buf;
      
      {
         char *p;

         for (p = buf; *p; p++){
	    if (w_curp[ch] < 0 || w_curp[ch] >= BLKSIZE){
	       if (write(w_fd[ch], w_buffer[ch], BLKSIZE) != BLKSIZE)
		  return(0);
	       w_curp[ch] = 0;
	       }
	    w_buffer[ch][w_curp[ch]++] = *p;
	    }
         return(1);
      }

/*
 * w_write - buffered write - write one character. check for a needed flush.
 */
   fb_w_write(ch, buf)
      int ch;
      char *buf;
      
      {
         if (w_curp[ch] < 0 || w_curp[ch] >= BLKSIZE){
	    if (write(w_fd[ch], w_buffer[ch], BLKSIZE) != BLKSIZE)
	       return(0);
	    w_curp[ch] = 0;
	    }
	 w_buffer[ch][w_curp[ch]++] = *buf;
         return(1);
      }

/*
 * w_writen - buffered write - write a block if needed.
 */
   fb_w_writen(ch, buf, n)
      int ch, n;
      char *buf;
      
      {
         char *p;

         for (p = buf; n > 0; p++, n--){
	    if (w_curp[ch] < 0 || w_curp[ch] >= BLKSIZE){
	       if (write(w_fd[ch], w_buffer[ch], BLKSIZE) != BLKSIZE)
		  return(0);
	       w_curp[ch] = 0;
	       }
	    w_buffer[ch][w_curp[ch]++] = *p;
	    }
         return(1);
      }

/*
 * wflush - flush anything left in the wbuffer
 */
   void fb_wflush(n)
      int n;
      
      {

         int i;

         for (i = 0; i < n; i++){
	    if (w_curp[i] > 0){
	       if (write(w_fd[i], w_buffer[i], w_curp[i]) != w_curp[i])
		  return;
	       w_curp[i] = 0;
	       }
	    }
      }
